lib/src/puppetdb-cli.cc

```diff
@@ -116,28 +197,132 @@ size_t write_body(char *ptr, size_t size, size_t nmemb, void *userdata){
     return size * nmemb;
 }
 
+// SynchronizedCharQueueWrapper is a rapidjson adapter which satisfies the
+// `ReadStream` interface for a rapidjson::Reader. This allows us to parse data
+// from curl as we recieve it.
+class SynchronizedCharQueueWrapper {
+  public:
+    typedef char Ch;
+    SynchronizedCharQueueWrapper(SynchronizedCharQueue& stream) : stream_(stream),
+                                                                  count_(0),
+                                                                  buffer_(stream_.pop()),
+                                                                  buffer_iterator_(buffer_.begin()) {}
+    char Peek() const {
+        int c = Front();
+        return c == char_traits<char>::eof() ? '\0' : static_cast<char>(c);
+    }
+    char Take() {
+        int c = Get();
+        return c == char_traits<char>::eof() ? '\0' : static_cast<char>(c);
+    }
+    size_t Tell() const { return count_; }
+    char* PutBegin() { assert(false); return 0; }
+    void Put(char c) { assert(false); }
+    void Flush() { assert(false); }
+    size_t PutEnd(char* c) { assert(false); return 0; }
+
+  private:
+    int Front() const { return *buffer_iterator_; }
+    int Get() {
+        int c = *buffer_iterator_;
```

Comment:
And not saying we have to do it, but at least in the C world, idiomatically, this would be "*buffer_iterator_++"...  Hence the old strcpy saw:
```C
  while(*dest++ = *src++);
```